### PR TITLE
Update tx structure

### PR DESCRIPTION
**Pull request by Sajjon**
C version of [JS PR](https://github.com/radixdlt/radixdlt-javascript/pull/154)

All unit tests have been migrated and are passing.

Successfully run one integration test from Javascript lib, the tx blob:

```
0a0001040c7e6ad291944d3fdf50cd278651e4d20ad28536b529004008a4c3938dce092c00000001092100000000000000000000000000000000000000000000000abbade0b6b3a76400000105000402935deebcad35bcf27d05b431276be8fcba26312cd1d54c33ac6748a72fe427ca0100000000000000000000000000000000000000000000d38b4d5d456f911400000005000000000105000402935deebcad35bcf27d05b431276be8fcba26312cd1d54c33ac6748a72fe427ca0100000000000000000000000000000000000000000000d38075ce8914caf400000eaf99885ac063393a2849d4b0df36c5ec3164408132526caf59f53d1239be2bf8000000000106000356959464545aa2787984fe4ac76496721a22f150c0076724ad7190fe3a597bb70402935deebcad35bcf27d05b431276be8fcba26312cd1d54c33ac6748a72fe427ca00000000000000000000000000000000000000000000000ad78ebc5ac6200000000921010000000000000000000000000000000000000000000000000de0b6b3a76400000105000402935deebcad35bcf27d05b431276be8fcba26312cd1d54c33ac6748a72fe427ca010000000000000000000000000000000000000000000000000de0b6b3a764000000
```

Which contains 3 transfers and one stake.
 
(this is the same blob used in `txParser.test.ts` test named `token_transfer_and_stake`).

### CODE REVIEW

File: src/transaction/transaction_parser.c [hunk 1]
```diff
@@ -46,11 +46,97 @@ bool parse_tx_fee_from_syscall(re_ins_syscall_t *syscall, uint256_t *tx_fee) {
...
+ * END
+ * SIG <signature>
+ *
+ * The corresponding fee paid in the transaction is then:
+ * `fee_paid = put_amt - sum(take_amt*)`
+ *
+ * There should only be one and only one `put_amt` value. There may be 0 or more `take_amt` values.
+ *
+ */
+static bool update_tx_fee_and_total_xrd_cost_if_needed(re_ins_syscall_t *syscall,
```
Comment (Sajjon) [inline src/transaction/transaction_parser.c:81]: @siy this function is new and probably the most notable and important update of this entire PR. It changes how tx fee is calculated.

**Comment by Sajjon** [comment]
Status: I think implementation wise it is ready to be merged.

I would probably test with some "real world" transactions first. I.e. wait for when this version can be used with Radix Wallet.

@siy so make a very temporary build, `bin`and `debug` folders zipped and send over Slack or something to @AbstractFruitFactory and maybe Caleb at Township so they can test it. Or of they give you a Radiw Wallet which points to a network supporting this new TX structure. Would be great to test those real world transactions before merging this PR.

Good luck!

File: src/instruction/instruction.c [hunk 2]
```diff
@@ -168,11 +193,13 @@ static bool does_substate_need_to_be_displayed(substate_t *substate, public_key_
             return true;
         case SUBSTATE_TYPE_PREPARED_UNSTAKE:
             return true;
-        case SUBSTATE_TYPE_STAKE_SHARE:
+        case SUBSTATE_TYPE_STAKE_OWNERSHIP:
+        case SUBSTATE_TYPE_VALIDATOR_ALLOW_DELEGATION_FLAG:
+        case SUBSTATE_TYPE_VALIDATOR_OWNER_COPY:
             return false;
+        default:
+            return false;  // should never happen
```
Comment (talekhinezh) [inline src/instruction/instruction.c:223]: Can we exit() with error rather than failing silently in these cases?

**Review by iamyulong — APPROVED** [review]
LGTM

File: src/instruction/instruction.h [hunk 3]
```diff
@@ -29,11 +35,18 @@ typedef struct {
     substate_t substate;
 } re_ins_up_t;
 
+/**
+ * Structure the RE instruction `INS_VREAD.
+ */
+typedef struct {
+    substate_t substate;
```
Comment (talekhinezh) [inline src/instruction/instruction.h]: vreads are now encoded differently from before. They are a different structure compared to substates and in general will only contain a key (for example only the validator key for AllowDelegationFlag).

File: src/transaction/transaction_parser.c [hunk 4]
```diff
@@ -46,11 +46,97 @@ bool parse_tx_fee_from_syscall(re_ins_syscall_t *syscall, uint256_t *tx_fee) {
...
+            return false;
+        }
+
+        // Copy over tx fee.
+        copy256(&transaction->tx_fee, &amount);
+
+        // Add tx fee to total cost
+        add256(&amount,
+               &transaction->total_xrd_amount_incl_fee,
+               &transaction->total_xrd_amount_incl_fee);
```
Comment (talekhinezh) [inline src/transaction/transaction_parser.c]: Looks very scary having the same pointer to source and target. Can we confirm for sure that in place updates will work for all values?

File: src/transaction/transaction_parser.c [hunk 5]
```diff
@@ -46,11 +46,97 @@ bool parse_tx_fee_from_syscall(re_ins_syscall_t *syscall, uint256_t *tx_fee) {
...
+
+        if (gt256(&amount, &transaction->tx_fee) ||
+            gt256(&amount, &transaction->total_xrd_amount_incl_fee)) {
+            // Invalid state, just parsed a `TX_FEE_RESERVE_TAKE` which was larger than
+            // amount left from put. This would result in a negative tx fee.
+            return false;
+        }
+
+        // TX_FEE = TX_FEE - TX_FEE_RESERVE_TAKE
+        minus256(&transaction->tx_fee, &amount, &transaction->tx_fee);
```
Comment (talekhinezh) [inline src/transaction/transaction_parser.c]: Same comment regarding in place updates.

File: src/transaction/transaction_parser.c [hunk 6]
```diff
@@ -46,11 +46,97 @@ bool parse_tx_fee_from_syscall(re_ins_syscall_t *syscall, uint256_t *tx_fee) {
...
+
+        // TOTAL_XRD_COST = TOTAL_XRD_COST - TX_FEE_RESERVE_TAKE
+        minus256(&transaction->total_xrd_amount_incl_fee,
+                 &amount,
+                 &transaction->total_xrd_amount_incl_fee);
```
Comment (talekhinezh) [inline src/transaction/transaction_parser.c]: More in place updates

File: src/types/uint256.c [hunk 7]
```diff
@@ -476,4 +476,10 @@ bool to_string_uint256_get_len(uint256_t *uint256,
 bool to_string_uint256(uint256_t *uint256, char *out, const size_t out_len) {
     size_t discarded = out_len;
     return to_string_uint256_get_len(uint256, out, out_len, &discarded);
+}
+
+bool is_uint256_greather_than_zero(uint256_t *target) {
+    uint256_t zero;
+    clear256(&zero);
```
Comment (talekhinezh) [inline src/types/uint256.c]: surely we can do better than using zero buffers :)

**Review by talekhinezh — CHANGES_REQUESTED** [review]

**Pull request merged**